src/net_processing.cpp

```diff
@@ -3214,6 +3243,377 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
+        pfrom->m_recon_state->incoming_recon = CNode::ReconPhase::BISEC_REQUESTED;
+        return true;
+    }
+
+    // Privacy leak can't happen here, because we will consider only those
+    // transactions which were sketched (stored in local set snapshot).
+    if (strCommand == NetMsgType::RECONCILDIFF) {
+        uint8_t success_value;
+        std::vector<uint32_t> ask_shortids;
+        vRecv >> success_value >> ask_shortids;
+        ReconResult success = static_cast<ReconResult>(success_value);
+        assert(pfrom->m_recon_state->incoming_recon == CNode::ReconPhase::INIT_RESPONDED || pfrom->m_recon_state->incoming_recon == CNode::ReconPhase::BISEC_RESPONDED);
+        std::vector<uint128> remote_missing;
+        if (success == RECON_SUCCESS) {
+            remote_missing = pfrom->GetTruncIDsFromShortIDs(ask_shortids, connman);
+        } else if (success == RECON_FAILED) {
+            assert(ask_shortids.size() == 0);
+            for (uint256 wtxid: pfrom->m_recon_state->local_set_snapshot) {
+                remote_missing.push_back(connman->GetTruncIDFromWTXID(wtxid));
+            }
+        } else {
+            assert(pfrom->m_recon_state->incoming_recon == CNode::ReconPhase::BISEC_RESPONDED);
+            for (uint256 wtxid: pfrom->m_recon_state->local_set_snapshot) {
+                uint32_t short_id = pfrom->ComputeShortID(wtxid);
+                // No matter which chunk failed, always announce what is asked for by short id
+                if (std::find(ask_shortids.begin(), ask_shortids.end(), short_id) != ask_shortids.end()) {
+                    remote_missing.push_back(connman->GetTruncIDFromWTXID(wtxid));
+                }
+                if (success == RECON_LOW_FAILED) {
+                    // High chunk bisection succeeded and asked for in ask_shortid, announce only remaining from the low chunk.
+                    if (short_id <= BISECTION_MEDIAN) {
+                        remote_missing.push_back(connman->GetTruncIDFromWTXID(wtxid));
+                    }
+                } else if (success == RECON_HIGH_FAILED) { 
+                    // Low chunk bisection succeeded and asked for in ask_shortid, announce only remaining from the high chunk
+                    if (short_id > BISECTION_MEDIAN) {
+                        remote_missing.push_back(connman->GetTruncIDFromWTXID(wtxid));
+                    }
+                }
+            }
+        }
+        if (remote_missing.size() != 0)
+            connman->PushMessage(pfrom, msgMaker.Make(NetMsgType::INVTX, remote_missing));
+        pfrom->FinalizeReconciliation(false, CNode::LocalQAction::Q_KEEP);
+        return true;
+    }
+
+    if (strCommand == NetMsgType::INVTX) {
+        std::vector<uint128> inv_truncids;
+        vRecv >> inv_truncids;
+        for (uint128 truncid: inv_truncids) {
+            // TODO
+            // Need to mimic INV processing, so re-implement AlreadyHave,
```

Comment:
AlreadyHave depends on global state. Instead of duplicating the global state, what about making AlreadyHave work on truncated wtxids?